In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
data=pd.read_csv("https://raw.githubusercontent.com/huynhthanh98/ML/master/lab-03/heart.csv")
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

## Câu 1: Hãy xây dựng mô hình logistic regression bằng tất cả các features trong file heart, so sánh với thư viện sklearn.

In [3]:
X=data.drop(columns=['target'])
y=data['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=5)
y_train=y_train.values.reshape([-1,1])
y_test=y_test.values.reshape([-1,1])

In [4]:
# Normalization
min=np.min(X_train,axis=0)
max=np.max(X_train,axis=0)
X_train=(X_train-min)/(max -min)

X_test=(X_test-min)/(max-min)

In [5]:
#Hàm sigmoid
def sigmoid(W,X):
  """
  W là trọng số
  X là sample(s)
  """

  return 1/(1+np.exp((-np.matmul(X,W))))

In [6]:
#Thêm 1 cột những số 1 vào X_train
X_train_bar=np.concatenate([np.ones([X_train.shape[0],1]),X_train],axis=1)
X_train_bar.shape

(242, 14)

In [7]:
lr=0.0001
#W_init=np.random.rand(14,1)
#W_init=np.array([[ 2.23789173],
#       [-0.96136103]]) #Đây là khởi tạo

W_init=np.zeros([X_train_bar.shape[1],1])
"""W_init=np.array([[ 0.38652764],
       [-1.0147613 ]]) #lưu ý đây là W_init gần với W cần tìm để chạy cho nhanh
       """
X_train_bar=np.concatenate([np.ones([X_train.shape[0],1]),X_train],axis=1)

list_W=[W_init]
list_error=[10]
for epoch in range(50000):
    W=list_W[-1]
    prediction=sigmoid(W,X_train_bar) # Dự đoán

    error=-np.mean((y_train*np.log(prediction)+(1-y_train)*np.log(1-prediction)),axis=0) #Tính error
    
    gradient=np.mean((prediction-y_train)*X_train_bar,axis=0) #Tính Gradient
    
    gradient=gradient.reshape(-1,1)
    
    W=W-lr*gradient #cập nhật W
    
    list_W.append(W)
    
    list_error.append(error)
    
    if epoch % 1000==0:
        print(list_error[-1])

[0.69314718]
[0.6897915]
[0.68656639]
[0.68345523]
[0.68044413]
[0.67752155]
[0.67467783]
[0.67190492]
[0.66919605]
[0.66654558]
[0.66394874]
[0.66140151]
[0.6589005]
[0.65644283]
[0.65402604]
[0.65164802]
[0.64930696]
[0.64700128]
[0.64472962]
[0.64249077]
[0.64028367]
[0.63810736]
[0.635961]
[0.63384382]
[0.63175512]
[0.62969424]
[0.62766059]
[0.62565362]
[0.62367279]
[0.62171763]
[0.61978766]
[0.61788242]
[0.61600151]
[0.61414449]
[0.61231098]
[0.61050059]
[0.60871295]
[0.6069477]
[0.60520447]
[0.60348293]
[0.60178273]
[0.60010355]
[0.59844506]
[0.59680695]
[0.5951889]
[0.59359061]
[0.59201177]
[0.5904521]
[0.58891129]
[0.58738908]


In [8]:
# Kết quả tự xây dựng
X_test_bar=np.concatenate([np.ones([X_test.shape[0],1]),X_test],axis=1)
a=sigmoid(W,X_test_bar)
np.mean(np.where(a>0.5,1,0)==y_test)

0.8032786885245902

In [9]:
# So sánh với thư viện Sklearn
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(random_state=0).fit(X_train, y_train)
clf.score(X_test,y_test)


C:\Users\PhongTTP\Anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


0.8852459016393442

### Kết quả hàm thủ công: 0.8032786885245902
### Kết quả khi dùng hàm scikit-learn: 0.8852459016393442

## Câu 2: Hãy xây dựng mô hình softmax regression trên bộ Iris (nên Normalize data), so sánh với thư viện sklearn.


In [10]:
from sklearn import datasets

# import some data to play with
iris = datasets.load_iris()
X = iris.data
y = iris.target

In [11]:
# train_test_split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [12]:
# Normalization
max=np.max(X_train,axis=0)
min=np.min(X_train,axis=0)

X_train=(X_train-min)/(max-min)
X_test=(X_test-min)/(max-min)

In [13]:
y_train_onehot = np.zeros( (y_train.size, y_train.max() + 1),dtype=int)
y_train_onehot[np.arange(y_train.size), y_train.reshape(-1)] = 1

y_test_onehot = np.zeros( (y_test.size, y_test.max() + 1) ,dtype=int)
y_test_onehot[np.arange(y_test.size), y_test.reshape(-1)] = 1


X_train_bar=np.concatenate([np.ones([X_train.shape[0],1]),X_train],axis=1)
X_test_bar=np.concatenate([np.ones([X_test.shape[0],1]),X_test],axis=1)
X_test_bar.shape

(50, 5)

In [14]:
lr=0.005

W_init=np.zeros([X_train_bar.shape[1],1])
list_W=[W_init]
list_error=[0]
for epoch in range(200000):
    expi=np.exp(np.matmul(X_train_bar,list_W[-1])) #300 3
    expsum=np.sum(expi,axis=1)   #300, 
    soft1=expi/expsum.reshape(-1,1)  #300 1
    a=(-y_train_onehot+soft1)    #300,1
    W_temp=np.zeros([X_train_bar.shape[1],y_train_onehot.shape[1]])
    #W_temp
    for i in range(y_train_onehot.shape[1]):
        temp=i+1
        W_temp[:,i:temp]=np.mean(a[:,i:temp]*X_train_bar,axis=0).reshape(-1,1)

#cal error
    W=list_W[-1]-lr*W_temp
    
    for i in range(y_train_onehot.shape[1]):
        error=np.exp(np.matmul(X_train_bar,W))/np.sum(np.exp(np.matmul(X_train_bar,W)),axis=1).reshape(-1,1)
        error=-y_train_onehot*np.log(error)
        error=np.sum(np.mean(error,axis=0))
        
    list_W.append(W)
    list_error.append(error)
    
    if epoch % 10000==0:
        print(list_error[-1])

1.098198230715856
0.4541107820263249
0.358430626617942
0.3099779574490078
0.27806815869501905
0.2544922052355366
0.23596141090505504
0.2208346998764794
0.20816830722140847
0.1973641532118171
0.18801639630607792
0.17983568437281305
0.17260799287018203
0.1661704904790502
0.16039650510625572
0.15518569451129904
0.15045736498087456
0.14614578431870937
0.1421968075643133
0.13856539472123702


In [15]:
expi=np.exp(np.matmul(X_test_bar,list_W[-1])) 
expsum=np.sum(expi,axis=1)   
#Kết quả
np.mean(y_test.reshape(-1)==np.argmax(expi/expsum.reshape(-1,1),axis=1))

0.98

In [16]:
# So sánh với thư viện SKLearn
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(random_state=0).fit(X_train, y_train)
clf.score(X_test,y_test)

0.9

### Kết quả hàm thủ công: 0.98
### Kết quả khi dùng hàm scikit-learn: 0.9